### PR TITLE
Add Single Class Model PR Curves

### PR DESCRIPTION
**Pull request by roshankern**
In this pull request, precision-recall curves are added for the single-class models trained in #25. A PR curve is made for each label type (eg positive and negative label), for each model (Large single-class model, Metaphase single-class model, etc), for each evaluation dataset (train or test), for each feature type (CP, DP, CP_and_DP). 

These PR curve figures have been added to [class_PR_curves.ipynb](https://github.com/roshankern/phenotypic_profiling_model/blob/evaluate-single-class-models/3.evaluate_model/class_PR_curves.ipynb) and the PR data have been saved to [precision_recall_curves](https://github.com/roshankern/phenotypic_profiling_model/tree/evaluate-single-class-models/3.evaluate_model/evaluations/precision_recall_curves).

### CODE REVIEW

**Comment by roshankern** [comment]
Unfortunately, during #25 (training single class models) all notebooks were run in my base `python` environment instead of the correct `phenotypic_profiling` environment. Thus, I reran those notebooks in the correct environment here as well. Please ignore all the recreated model files (ending in `.joblib`) by scrolling to the bottom of the pull request review.

**Comment by roshankern** [comment]
This PR is ready for review!

File: 3.evaluate_model/README.md [hunk 1]
```diff
@@ -15,7 +15,9 @@ F1 score measures the models precision and recall performance for each phenotypi
 The f1 score data are saved to [F1_scores](evaluations/F1_scores).
 
 In [class_PR_curves.ipynb](class_PR_curves.ipynb), we use [sklearn.metrics.precision_recall_curve](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html) to derive the precision-recall curves for each model, feature type, and dataset combination.
-The precision recall curves and their data are saved to [class_precision_recall_curves](evaluations/class_precision_recall_curves/).
+These PR curves are created for each label type of the logistic regression model.
+E.g. each multi-class model has 15 labels (1 for each phentoypic class) and 15 PR curves while each single-class model has 2 labels (positive and negative label for its repective phenotype) and 2 PR curves.
```
Comment (d33bs) [inline 3.evaluate_model/README.md]:
```suggestion
E.g. each multi-class model has 15 labels (1 for each phenotypic class) and 15 PR curves while each single-class model has 2 labels (positive and negative label for its respective phenotype) and 2 PR curves.
```

File: 3.evaluate_model/README.md [hunk 2]
```diff
@@ -30,7 +32,9 @@ The LOIO evaluation procedure is as follows:
 
 The probabilities are saved to [LOIO_probas](evaluations/LOIO_probas).
 
-**Note:** Intermediate `.tsv` data are stored in tidy format, a standardized data structure (see [Tidy Data](https://vita.had.co.nz/papers/tidy-data.pdf) by Hadley Wickham for more details).
+**Notes:** 
+1) Intermediate `.tsv` data are stored in tidy format, a standardized data structure (see [Tidy Data](https://vita.had.co.nz/papers/tidy-data.pdf) by Hadley Wickham for more details).
+2) SCM stands for "single cell model(s)" and is used as an abbrevation for the binary, sinlge-class models throughout this module.
```
Comment (d33bs) [inline 3.evaluate_model/README.md:37]: Really appreciate the dedication to audience understanding with this line!

File: 3.evaluate_model/class_PR_curves.ipynb
Comment (d33bs) [inline 3.evaluate_model/class_PR_curves.ipynb:1]: I noticed this file didn't have any output for the cells. Double checking: was that intentional, and if not, should it be updated to include the results?

File: 3.evaluate_model/scripts/nbconverted/class_PR_curves.py [hunk 3]
```diff
@@ -88,27 +83,135 @@
...
+phenotypic_classes = features_dataframe["Mitocheck_Phenotypic_Class"].unique()
+
+# colors for plotted lines
+data_split_colors = {
+    "CP_and_DP (test)": "#C214CB",
+    "CP_and_DP (train)": "#E88EED",
+    "CP (test)": "#EB4B4B",
+    "CP (train)": "#F8B5B5",
+    "DP (test)": "#5158bb",
+    "DP (train)": "#B5B9EA",
```
Comment (d33bs) [inline 3.evaluate_model/scripts/nbconverted/class_PR_curves.py:139]: Would it provide any benefit to consider the use of [colormaps](https://matplotlib.org/stable/tutorials/colors/colormaps.html) somehow for generating these colors?

File: 3.evaluate_model/scripts/nbconverted/class_PR_curves.py [hunk 4]
```diff
@@ -88,27 +83,135 @@
...
-# ### Save PR curves from each evaluation
+# ### Save PR curves from each evaluation (multiclass)
 # 
 
 # In[4]:
 
 
 # compile list of tidy data into one dataframe
-compiled_class_PR_curves = pd.concat(compiled_class_PR_curves).reset_index(drop=True)
+# some thresholds are None because last PR value doesnt correspond to cell dataset (these values are always P=1, R=0), remove these rows from PR data
```
Comment (d33bs) [inline 3.evaluate_model/scripts/nbconverted/class_PR_curves.py]:
```suggestion
# some thresholds are None because last PR value doesn't correspond to cell dataset (these values are always P=1, R=0), remove these rows from PR data
```

File: 3.evaluate_model/scripts/nbconverted/class_PR_curves.py [hunk 5]
```diff
@@ -88,27 +83,135 @@
...
+
+
+# ### Save PR curves from each evaluation (multiclass)
+# 
+
+# In[6]:
+
+
+# compile tidy PR data
+# some thresholds are None because last PR value doesnt correspond to cell dataset (these values are always P=1, R=0), remove these rows from PR data
```
Comment (d33bs) [inline 3.evaluate_model/scripts/nbconverted/class_PR_curves.py]:
```suggestion
# some thresholds are None because last PR value doesn't correspond to cell dataset (these values are always P=1, R=0), remove these rows from PR data
```

File: utils/evaluate_utils.py [hunk 6]
```diff
@@ -99,6 +99,149 @@ def class_PR_curves(
...
+    # only label outer plots
+    for ax in axs.flat:
+        ax.label_outer()
+
+    # add legend to figure with all subplots
+    handles, labels = ax.get_legend_handles_labels()
+    fig.legend(handles, labels, loc="upper right")
+
+    # compile PR data
+    # some thresholds are None because last PR value doesnt correspond to cell dataset (these values are always P=1, R=0), remove these rows from PR data
```
Comment (d33bs) [inline utils/evaluate_utils.py]:
```suggestion
    # some thresholds are None because last PR value doesn't correspond to cell dataset (these values are always P=1, R=0), remove these rows from PR data
```

File: utils/evaluate_utils.py [hunk 7]
```diff
@@ -99,6 +99,149 @@ def class_PR_curves(
...
 
+def class_PR_curves_SCM(
+    single_cell_data: pd.DataFrame,
+    single_class_model: LogisticRegression,
+    fig: Figure,
+    axs: np.ndarray,
+    phenotypic_class_index: int,
+    data_split_colors: dict,
+    model_type: str,
+    feature_type: str,
```
Comment (d33bs) [inline utils/evaluate_utils.py]: Based on what's found in the docstring below, consider using [`typing.Literal`](https://docs.python.org/3.8/library/typing.html#typing.Literal) here to specify certain static strings (instead of any string).

File: utils/evaluate_utils.py [hunk 8]
```diff
@@ -99,6 +99,149 @@ def class_PR_curves(
...
+        feature type model uses (CP, DP, or CP_and_DP)
+    evaluation_type : str
+        type of data being used for evaluation (test or train)
+    phenotypic_class : str
+        phenotypic class of single cell model being evaluated
+
+    Returns
+    -------
+    pd.DataFrame
+        _description_
```
Comment (d33bs) [inline utils/evaluate_utils.py]: Consider filling out the details here for the return data.

**Review by d33bs — APPROVED** [review]
Nice work! I left a few comments with this review and overall thought it LGTM! Please don't hesitate to let me know if you have any questions or if I may clarify at all.

File: 3.evaluate_model/scripts/nbconverted/class_PR_curves.py (hunk 3 above)
Comment (roshankern) [reply 3.evaluate_model/scripts/nbconverted/class_PR_curves.py:139]: I think in the future this would be worth implementing. For now I think it's best to keep these colors the same as those established by gwaybio in #24

File: 3.evaluate_model/class_PR_curves.ipynb
Comment (roshankern) [reply 3.evaluate_model/class_PR_curves.ipynb:1]: Good catch! This was supposed to have outputs. These have been added in https://github.com/WayScience/phenotypic_profiling_model/pull/26/commits/987e1ec5a50e2751ae962ec53d06265164fcb301

File: 3.evaluate_model/scripts/nbconverted/class_PR_curves.py (hunk 4 above)
Comment (roshankern) [reply 3.evaluate_model/scripts/nbconverted/class_PR_curves.py]: Added to notebook in https://github.com/WayScience/phenotypic_profiling_model/pull/26/commits/987e1ec5a50e2751ae962ec53d06265164fcb301

File: 3.evaluate_model/scripts/nbconverted/class_PR_curves.py (hunk 5 above)
Comment (roshankern) [reply 3.evaluate_model/scripts/nbconverted/class_PR_curves.py]: Added to notebook in https://github.com/WayScience/phenotypic_profiling_model/commit/987e1ec5a50e2751ae962ec53d06265164fcb301

File: utils/evaluate_utils.py (hunk 7 above)
Comment (roshankern) [reply utils/evaluate_utils.py]: I've never seen this before, thanks for teaching me something new! Added in https://github.com/WayScience/phenotypic_profiling_model/commit/987e1ec5a50e2751ae962ec53d06265164fcb301

File: utils/evaluate_utils.py (hunk 6 above)
Comment (roshankern) [reply utils/evaluate_utils.py]: Added to notebook in https://github.com/WayScience/phenotypic_profiling_model/commit/987e1ec5a50e2751ae962ec53d06265164fcb301

File: utils/evaluate_utils.py (hunk 8 above)
Comment (roshankern) [reply utils/evaluate_utils.py]: Great catch! Fixed in https://github.com/WayScience/phenotypic_profiling_model/commit/987e1ec5a50e2751ae962ec53d06265164fcb301
